hawkular-alerts-engine-extensions/hawkular-alerts-aggregated-events-alerter/src/main/java/org/hawkular/alerts/extensions/CepEngineImpl.java

```diff
@@ -0,0 +1,203 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.alerts.extensions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+import javax.annotation.Resource;
+import javax.ejb.EJB;
+import javax.ejb.Lock;
+import javax.ejb.LockType;
+import javax.ejb.Singleton;
+import javax.ejb.Startup;
+import javax.enterprise.concurrent.ManagedExecutorService;
+
+import org.hawkular.alerts.api.model.event.Event;
+import org.hawkular.alerts.api.model.trigger.FullTrigger;
+import org.hawkular.alerts.api.services.AlertsService;
+import org.jboss.logging.Logger;
+import org.kie.api.KieBase;
+import org.kie.api.KieBaseConfiguration;
+import org.kie.api.conf.EventProcessingOption;
+import org.kie.api.event.rule.AfterMatchFiredEvent;
+import org.kie.api.event.rule.AgendaEventListener;
+import org.kie.api.event.rule.AgendaGroupPoppedEvent;
+import org.kie.api.event.rule.AgendaGroupPushedEvent;
+import org.kie.api.event.rule.BeforeMatchFiredEvent;
+import org.kie.api.event.rule.MatchCancelledEvent;
+import org.kie.api.event.rule.MatchCreatedEvent;
+import org.kie.api.event.rule.ObjectDeletedEvent;
+import org.kie.api.event.rule.ObjectInsertedEvent;
+import org.kie.api.event.rule.ObjectUpdatedEvent;
+import org.kie.api.event.rule.RuleFlowGroupActivatedEvent;
+import org.kie.api.event.rule.RuleFlowGroupDeactivatedEvent;
+import org.kie.api.event.rule.RuleRuntimeEventListener;
+import org.kie.api.io.ResourceType;
+import org.kie.api.runtime.KieSession;
+import org.kie.api.runtime.KieSessionConfiguration;
+import org.kie.api.runtime.conf.ClockTypeOption;
+import org.kie.internal.KnowledgeBaseFactory;
+import org.kie.internal.utils.KieHelper;
+
+/**
+ * It evaluates events externally and send data into alerting
+ *
+ * @author Jay Shaughnessy
+ * @author Lucas Ponce
+ */
+@Singleton
+@Startup
+public class CepEngineImpl implements CepEngine {
+    private final Logger log = Logger.getLogger(CepEngineImpl.class);
+
+    Expression expression;
+    List<Event> results;
+    KieSession kieSession;
+
+    @EJB
+    private AlertsService alertsService;
+
+    @Resource
+    private ManagedExecutorService executor;
+
+    @Lock(LockType.READ)
+    public void processEvents(Collection<Event> events) {
+        if (kieSession != null) {
+            events.stream().forEach(e -> kieSession.insert(e));
```

Comment:
Indeed, the caller will use always a TreeSet<Event> as a parameter.
I don't think that it might have a future usecase with a different type, so perhaps here we should propagate the TreeSet as the input parameter.